### PR TITLE
Fix for Database blips

### LINKED ISSUE #337: Exception when Database blips
### Expected behaviour

When the database disconnects/reconnects operations resume normally
### Actual behaviour

There is an exception seen
### Steps to reproduce the behaviour

This occurred when changing the instance size of the RDS database; may occurs in other scenarios too, such as switching from master to slave, needs checking.
### Technical information

See attached stack trace:
[database_exception.txt](https://github.com/ONSdigital/eq-survey-runner/files/385769/database_exception.txt)

### PR DESCRIPTION
**Pull request by warrenbailey**
### What is the context of this PR?

Fixes #337 

Test database connections when they are requested from the pool, if they are invalid try them again as this should reinitialise the connection. If that still doesn't work then forcibly close them.
### How to review
1. Deploy this branch to elastic beanstalk
2. Restart the RDS instance
3. Check you can still complete a survey
- [ ] Do all commits have sensible commit messages?
- [ ] Is all new code unit tested?
- [ ] Are there integration tests if needed?
- [ ] Is there sufficient logging?
- [ ] Is there documentation or is the code self-describing?
### Who worked on the PR

@warrenbailey

…

### CODE REVIEW

File: app/data_model/database.py [hunk 1]
```diff
@@ -68,8 +71,32 @@ def create_session_and_engine():
...
 
 base.metadata.create_all(engine)
+
+
+@event.listens_for(engine, "engine_connect")
+def test_connection(connection, branch):
+    if branch:
+        # "branch" refers to a sub-connection of a connection,
+        # we don't want to bother testing on these.
+        return
```
Comment (dhilton) [inline app/data_model/database.py:85]: What does this return? If the function has finished perhaps return None or True? Further on the code returns errors, so be good to have a return value under any scenario.

Comment (warrenbailey) [inline app/data_model/database.py:85]: It's a listener function, it doesn't return anything. The idea being it allows you to execute some functionality when a connection is checked out of the pool. In this case we test the connection, if it fails tries to reconnect and if the reconnect fails it then forcibly closes the connection.

It's a modified version of:
http://docs.sqlalchemy.org/en/latest/core/pooling.html#disconnect-handling-pessimistic

File: app/main/views/root.py [hunk 2]
```diff
@@ -68,5 +68,5 @@ def login():
     except InvalidTokenException as e:
         logger.warning("Invalid Token provided")
         return errors.forbidden(e)
-    except RuntimeError as e:
+    except (Exception, RuntimeError) as e:
```
Comment (warrenbailey) [inline app/main/views/root.py:71]: Different file ;-)

File: app/data_model/database.py (hunk 1 above)
Comment (dhilton) [reply app/data_model/database.py:85]: Ah ok cool - 👍 from me then.

**Comment by warrenbailey** [comment]
Rebasing

**Comment by iwootten** [comment]
So, having spent the best part to the afternoon getting this up, this appears to fail for me when I restart RDS. 

I can get go through the motions of a survey initially, and successfully restart RDS. When I attempt another questionnaire, can get to the introduction, but on proceeding, get an internal server error.

**Comment by warrenbailey** [comment]
@iwootten I've tested this again this morning and it works for me. I do get one internal server error while the database is restarting but once it's up its all back to normal.

**Comment by iwootten** [comment]
Yep works for me today too. Guess I was too eager after restarting the db yesterday.

**Pull request merged**